src/lib/misc/roughtime/roughtime.cpp

```diff
@@ -0,0 +1,527 @@
+   const uint32_t start_content = num_tags * 8;
+   if(start_content > bytes.size())
+      { throw Invalid_Argument("invalid structure"); }
+   uint32_t start = start_content;
+   std::map<std::string, std::vector<uint8_t>> tags;
+   for(unsigned i=0; i<num_tags; ++i)
+      {
+      const uint32_t end = ((i+1) == num_tags) ? bytes.size() : start_content + *(uint32_t*)(buf + 4 + i*4);
+      if(end > bytes.size())
+         { throw Invalid_Argument("invalid structure"); }
+      if(end < start)
+         { throw Invalid_Argument("invalid structure"); }
+      const char* label_ptr = cast_uint8_ptr_to_char(buf) + (num_tags+i)*4;
+      const char label[] = {label_ptr[0], label_ptr[1], label_ptr[2], label_ptr[3], 0};
+      tags[label] = std::vector<uint8_t>(buf+start, buf+end);
+      start = end;
+      }
+   return tags;
+   }
+
+template<typename T>
+T get(const std::map<std::string, std::vector<uint8_t>>& map, const std::string& label)
+   {
+   const auto& tag = map.find(label);
+   if(tag == map.end())
+      { throw Invalid_Argument("tag not found"); }
+   if(tag->second.size() != sizeof(T))
+      { throw Invalid_Argument("tag content wrong size"); }
+   return *(const T*)tag->second.data();
+   }
+
+const std::vector<uint8_t>& get_v(const std::map<std::string, std::vector<uint8_t>>& map, const std::string& label)
+   {
+   const auto& tag = map.find(label);
+   if(tag == map.end())
+      { throw Invalid_Argument("tag not found"); }
+   return tag->second;
+   }
+
+bool verify_signature(const std::array<uint8_t, 32>& pk, const std::vector<uint8_t>& payload,
+                      const std::array<uint8_t, 64>& signature)
+   {
+   const char context[] = "RoughTime v1 response signature";
+   Ed25519_PublicKey key(std::vector<uint8_t>(pk.data(), pk.data()+pk.size()));
+   PK_Verifier verifier(key, "Pure");
+   verifier.update(cast_char_ptr_to_uint8(context), sizeof(context)); //add context including \0
+   verifier.update(payload);
+   return verifier.check_signature(signature.data(), signature.size());
+   }
+
+std::array<uint8_t, 64> hashLeaf(const std::array<uint8_t, 64>& leaf)
+   {
+   std::array<uint8_t, 64> ret;
+   std::unique_ptr<HashFunction> hash(HashFunction::create("SHA-512"));
+   hash->update(0);
+   hash->update(leaf.data(), leaf.size());
+   hash->final(ret.data());
+   return ret;
+   }
+
+std::array<uint8_t, 64> hashNode(const std::array<uint8_t, 64>& left, const std::array<uint8_t, 64>& right)
+   {
+   std::array<uint8_t, 64> ret;
+   std::unique_ptr<HashFunction> hash(HashFunction::create("SHA-512"));
+   hash->update(1);
+   hash->update(left.data(), left.size());
+   hash->update(right.data(), right.size());
+   hash->final(ret.data());
+   return ret;
+   }
+
+template<size_t N>
+std::array<uint8_t, N> vector_to_array(std::vector<uint8_t> vec)
+   {
+   if(vec.size() != N)
+      { throw Invalid_Argument("invalid vector size"); }
+   return *(std::array<uint8_t, N>*)vec.data();
```

Comment:
This case/conversion looks really strange to me. Is this a legit transformation? It seems like a violation of strict aliasing.

Reply:
It's a violation... I only understood it is unacceptable recently.